File: tests/abi/capability_state_layout.cpp
#include "pluginterfaces/base/funknown.h"
#include "pluginterfaces/vst/ivstautomationstate.h"
#include "pluginterfaces/vst/ivstpluginterfacesupport.h"
#include "pluginterfaces/vst/ivstprefetchablesupport.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("ePrefetchableSupport.kIsNeverPrefetchable %d\n", Steinberg::Vst::kIsNeverPrefetchable);
	std::printf ("ePrefetchableSupport.kIsYetPrefetchable %d\n", Steinberg::Vst::kIsYetPrefetchable);
	std::printf ("ePrefetchableSupport.kIsNotYetPrefetchable %d\n", Steinberg::Vst::kIsNotYetPrefetchable);
	std::printf ("ePrefetchableSupport.kNumPrefetchableSupport %d\n", Steinberg::Vst::kNumPrefetchableSupport);
	std::printf ("IAutomationState.kNoAutomation %d\n", Steinberg::Vst::IAutomationState::kNoAutomation);
	std::printf ("IAutomationState.kReadState %d\n", Steinberg::Vst::IAutomationState::kReadState);
	std::printf ("IAutomationState.kWriteState %d\n", Steinberg::Vst::IAutomationState::kWriteState);
	std::printf ("IAutomationState.kReadWriteState %d\n", Steinberg::Vst::IAutomationState::kReadWriteState);

	PRINT_TYPE (IPlugInterfaceSupport);
	PRINT_TYPE (IPrefetchableSupport);
	PRINT_TYPE (IAutomationState);

	print_iid ("IPlugInterfaceSupport", Steinberg::Vst::IPlugInterfaceSupport_iid);
	print_iid ("IPrefetchableSupport", Steinberg::Vst::IPrefetchableSupport_iid);
	print_iid ("IAutomationState", Steinberg::Vst::IAutomationState_iid);
	return 0;
}

File: tools/capability_state_layout.zig
const std = @import("std");
const automation = @import("zig-vst3").pluginterfaces.vst.ivstautomationstate;
const interface_support = @import("zig-vst3").pluginterfaces.vst.ivstpluginterfacesupport;
const prefetch = @import("zig-vst3").pluginterfaces.vst.ivstprefetchablesupport;

pub fn main(init: std.process.Init) !void {
    var stdout_buffer: [4096]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(init.io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    defer stdout.flush() catch {};
    try stdout.print("ePrefetchableSupport.kIsNeverPrefetchable {}\n", .{@intFromEnum(prefetch.ePrefetchableSupport.kIsNeverPrefetchable)});
    try stdout.print("ePrefetchableSupport.kIsYetPrefetchable {}\n", .{@intFromEnum(prefetch.ePrefetchableSupport.kIsYetPrefetchable)});
    try stdout.print("ePrefetchableSupport.kIsNotYetPrefetchable {}\n", .{@intFromEnum(prefetch.ePrefetchableSupport.kIsNotYetPrefetchable)});
    try stdout.print("ePrefetchableSupport.kNumPrefetchableSupport {}\n", .{@intFromEnum(prefetch.ePrefetchableSupport.kNumPrefetchableSupport)});
    try stdout.print("IAutomationState.kNoAutomation {}\n", .{automation.AutomationStates.kNoAutomation});
    try stdout.print("IAutomationState.kReadState {}\n", .{automation.AutomationStates.kReadState});
    try stdout.print("IAutomationState.kWriteState {}\n", .{automation.AutomationStates.kWriteState});
    try stdout.print("IAutomationState.kReadWriteState {}\n", .{automation.AutomationStates.kReadWriteState});

    try printType(stdout, "IPlugInterfaceSupport", interface_support.IPlugInterfaceSupport);
    try printType(stdout, "IPrefetchableSupport", prefetch.IPrefetchableSupport);
    try printType(stdout, "IAutomationState", automation.IAutomationState);

    try printTuid(stdout, "IPlugInterfaceSupport", interface_support.iplug_interface_support_iid);
    try printTuid(stdout, "IPrefetchableSupport", prefetch.iprefetchable_support_iid);
    try printTuid(stdout, "IAutomationState", automation.iautomation_state_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
